### ISSUE TITLE
Regarding visualization and router

### ISSUE THREAD

**Issue by yinangit** [UNITES-Lab/flex-moe#3]
Great work !

1. How were the visualizations of Figures 4 and 5 made? Taking Figure 4 as an example, how is the missing modality bank with a shape of [$`2^{|M|}-1`$, num_modality, num_patch, hidden_dim] used to visualize cosine similarity?
2. Is the difference between G-Router and S-Router only based on whether the proposed loss $`L_{ce}`$ is used or not? Does G-Router only use $`L_{balance}`$, while S-Router uses $`L_{ce}`$ and $`L_{balance}`$? And their structures are the same, just using different loss functions during different training periods?

Looking forward to your answer !

**Comment by benjamin-burns** [comment]
Hi @yinangit, I had similar questions about the precise definitions of the G-Router and S-Router. After carefully looking through the code, I can summarize the differences for you and other interested readers.

The "G-Router" and "S-Router" are implemented as a single class, `AddtionalNoisyGate` [sic] in `moe_module.py` rather than as separate classes. The behavior of `AddtionalNoisyGate` is dictated by the input batch, sometimes acting as the "G-Router" and other times acting as the "S-Router."

### G-Router

The G-Router is active when all observations in the given batch have no missing modalities. The G-Router loss is essentially $L_{balance}$ as described in the paper, with two clarifications. First, the given load loss
```math
\text{load}_e = \sum_{i}^{N}\delta (g_{ie}>0)
```
is the _evaluation_ load loss. The training load loss used by the authors is a different formula implemented in `FMoE`'s `NoisyGate`. Second, since the G-Router is only active when all observations in a batch have no missing modalities, the importance loss and load loss are calculated over _all_ experts, including the top-1 expert.

### S-Router

The S-Router is active when at least one observation in the given batch is missing modalities. The S-Router loss is essentially $L_{balance} + L_{ce}$, again with a few clarifications. 

- The MC notation in $L_{ce}$ is a potential cause for confusion, since it implies that the modality combination label, as defined in the paper, is the target distribution. This is nonsensical, and we should take MC($x_i$) here to be a one-hot vector encoding of the modality combination label. 
- $L_{ce}$ is not computed over all observations in the batch, as the formula in the paper would imply. Instead, $L_{ce}$ is only computed for observations that have missing modalities. As the code is currently presented, I don't believe there is any direct supervision encouraging full-modality observations to be routed to the "full-modality" expert.
- The gating balance loss is computed separately for full-modality observations and observations missing modalities. These two losses are summed to get $L_{balance}$. Again, the provided $\text{load}_e$ is for evaluation; the training version is omitted.

If I have gotten anything wrong, the authors should feel free to correct me.

**Comment by Chrisa142857** [comment]
@benjamin-burns Thank you for the clear description. Any idea of the first question by @yinangit? 

I think this question is actually how to implement $\mathcal B$, the Missing Modality Bank in the paper. Fig 4 and 5 look like the Pearson correlationship respectively in the first two dimensions of the matrix $\mathcal B$, which should be a tensor with shape [num_combination, num_modality, num_patch, hidden_dim]. Although I didn't find codes leveraging $\mathcal B$, I assume it is actually the MoE output `moe_outp` by `class FixedFMoE` in `moe_module.py`. 

Unclear points:
 - **(1)** is how to deal with the last two dimensions of $\mathcal B$. It could be averaging across each value in the last two dim. 
 - **(2)** is how Eq (2) in the paper is implemented.

**Comment by benjamin-burns** [comment]
The missing modality bank is implemented as a "standalone" PyTorch nn.Parameter object. There is no custom class that extends nn.Module. It is initialized in `main.py` on line 124 and used in the forward pass on line 84.

I find Figure 4 confusing as the authors do not make it clear how the missing modality bank embeddings are aggregated for the pairwise comparisons. Take the 0.63 cosine similarity between ICBG and C, for example. Is the vector for C the average all the missing modality bank I patches, B patches, and G patches when only C is available? This ***greatly*** influences how we should interpret these similarities!

You are probably correct @Chrisa142857 in assuming that they average along the num_patches dimension and either the observed modalities dimension (left plot) or missing modality dimension (right plot). We can't be sure without the code that creates Figure 4.
